File: symfony/src/Service/Cards/DeckTagService.php
<?php

namespace App\Service\Cards;

use App\Entity\Cards\DeckTag;
use App\Repository\Cards\DeckTagRepository;

class DeckTagService
{
    public function __construct(
        private DeckTagRepository $repository,
    ) {}

    public function create(array $data): DeckTag
    {
        throw new \LogicException('Not implemented');
    }

    public function update(DeckTag $entity, array $data): DeckTag
    {
        throw new \LogicException('Not implemented');
    }

    public function rename(int $id, $newName): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('DeckTag not found: ' . $id);
        $entity->rename($newName);
        $this->repository->save($entity, flush: true);
    }

    public function mergeInto(int $id, $targetTagId): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('DeckTag not found: ' . $id);
        $entity->mergeInto($targetTagId);
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/cards/deck_tag_service.rb
module Cards
  class DeckTagService
    def initialize(repository = DeckTag)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def rename(id, new_name)
      instance = DeckTag.find(id)
      instance.rename(new_name)
      instance.save!
    end

    def merge_into(id, target_tag_id)
      instance = DeckTag.find(id)
      instance.merge_into(target_tag_id)
      instance.save!
    end
  end
end
